### PR TITLE
[SPARK-14445][SQL] Support native execution of SHOW COLUMNS and SHOW PARTITIONS

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?

This PR adds Native execution of SHOW COLUMNS and SHOW PARTITION commands.

Command Syntax:

``` SQL
SHOW COLUMNS (FROM | IN) table_identifier [(FROM | IN) database]
```

``` SQL
SHOW PARTITIONS [db_name.]table_name [PARTITION(partition_spec)]
```
## How was this patch tested?

Added test cases in HiveCommandSuite to verify execution and DDLCommandSuite
to verify plans.

### CODE REVIEW

**Comment by dilipbiswal** [comment]
Currently i throw an exception for data source tables as in my understanding the
partition spec metadata is not available in the hive meta store. I saw a [PR-12204] (https://github.com/apache/spark/pull/12204) from @viirya. If with this we can have the
partitioning metadata from meta store, then we can remove this restriction.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @yhuai @andrewor14 @hvanhovell Could you review this PR too? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 1]
```diff
@@ -423,6 +424,100 @@ case class ShowTablePropertiesCommand(
...
+ * This function creates a [[ShowPartitionsCommand]] logical plan
+ *
+ * The syntax of using this command in SQL is:
+ * {{{
+ *   SHOW PARTITIONS [db_name.]table_name [PARTITION(partition_spec)]
+ * }}}
+ */
+case class ShowPartitionsCommand(
+    table: TableIdentifier,
+    partitionSpec: Option[Map[String, String]]) extends RunnableCommand {
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: please use `ExternalCatalog.TablePartitionSpec` here

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -404,6 +404,28 @@ private[hive] class HiveClientImpl(
     Option(hivePartition).map(fromHivePartition)
   }
 
+  /**
+   * Returns the partition names from hive metastore for a given table in a database.
+   */
+  override def getPartitionNames(
+      db: String,
+      table: String,
+      range: Short): Seq[String] = withHiveState {
```
Comment (andrewor14) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: what is this `range` thing? It seems that it's never used?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 3]
```diff
@@ -409,6 +409,25 @@ class SessionCatalog(
...
+   */
+  def getPartitionNames(db: String, table: String, range: Short): Seq[String] = {
+    externalCatalog.getPartitionNames(db, table, range)
+  }
+
+  /**
+   * Returns the partition names that matche the partition spec for a given table in a database.
+   * When no match is found, an empty Sequence is returned.
+   */
+  def getPartitionNames(
```
Comment (andrewor14) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: I think a better API is:
```
def listPartitions(
    db: String,
    table: String,
    partialSpec: Option[TablePartitionSpec]): Seq[CatalogTablePartition]
```

Note that we already have a `listPartitions`, but it doesn't allow us to match anything. Once we have this we can do the formatting ourselves in the `ShowPartitions` DDL. This leads to a more consistent catalog API and allows us to provide a less awkward implementation in `InMemoryCatalog`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 4]
```diff
@@ -423,6 +424,100 @@ case class ShowTablePropertiesCommand(
...
+  // The result of SHOW COLUMNS has one column called 'result'
+  override val output: Seq[Attribute] = {
+    AttributeReference("result", StringType, nullable = false)() :: Nil
+  }
+
+  override def run(sqlContext: SQLContext): Seq[Row] = {
+    val relation = sqlContext.sessionState.catalog.lookupRelation(table, None)
+    relation.schema.fields.map { field =>
+      Row(field.name)
+    }
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: It's better to get this information from the table metadata. 1 line:
```
sqlContext.sessionState.catalog.getTableMetadata(table).schema.map(_.name)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 5]
```diff
@@ -423,6 +424,100 @@ case class ShowTablePropertiesCommand(
...
+   * 3. If it is a view or index table.
+   */
+  private def checkRequirements(table: CatalogTable): Unit = {
+    if (table.tableType == CatalogTableType.VIRTUAL_VIEW ||
+      table.tableType == CatalogTableType.INDEX_TABLE) {
+      throw new AnalysisException("Operation not allowed: view or index table")
+    } else if (!DDLUtils.isTablePartitioned(table)) {
+      throw new AnalysisException(s"Table ${table.qualifiedName} is not a partitioned table")
+    } else if (DDLUtils.isDatasourceTable(table)) {
+      throw new AnalysisException("Operation not allowed: datasource table")
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: these error messages are too terse; "datasource table" is not an operation. I think it should sound more like:
```
SHOW PARTITIONS is not allowed on a view or index table: $name
SHOW PARTITIONS is not allowed on a table that is not partitioned: $name
SHOW PARTITIONS is not allowed on a datasource table: $name
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 6]
```diff
@@ -423,6 +424,100 @@ case class ShowTablePropertiesCommand(
...
+    }
+  }
+
+  override def run(sqlContext: SQLContext): Seq[Row] = {
+    val catalog = sqlContext.sessionState.catalog
+    val db = table.database.getOrElse(catalog.getCurrentDatabase)
+    if (catalog.isTemporaryTable(table)) {
+      Seq.empty[Row]
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: I think we should throw an exception instead if it's a temporary table

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 7]
```diff
@@ -423,6 +424,100 @@ case class ShowTablePropertiesCommand(
...
+    }
+  }
+
+  /**
+   * Validates and throws an [[AnalysisException]] exception under the following conditions:
+   * 1. If the table is not partitioned.
+   * 2. If it is a datasource table.
+   * 3. If it is a view or index table.
+   */
+  private def checkRequirements(table: CatalogTable): Unit = {
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: this is only called in 1 place. I would just inline it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 5 above)
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: also since these all throw exceptions anyway I would just do `if`s instead of `else if`s

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 8]
```diff
@@ -423,6 +424,100 @@ case class ShowTablePropertiesCommand(
...
+  }
+
+  /**
+   * This function validates the partitioning spec by making sure all the referenced columns are
+   * defined as partitioning columns in table definition. An AnalysisException exception is
+   * thrown if the partitioning spec is invalid.
+   */
+  private def validatePartitionSpec(table: CatalogTable, spec: Map[String, String]): Unit = {
+    if (!spec.keySet.forall(table.partitionColumns.map(_.name).contains)) {
+      throw new AnalysisException(s"Partition spec ${spec.mkString("(", ", ", ")")} contains " +
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: to provide a better error message I would do:
```
val badColumns = spec.keySet.filterNot(table.partitionColumnNames.contains)
throw new AnalysisException("Non-partitioned column(s) ${badColumns.mkString(", ")} are specified for SHOW PARTITIONS")
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 9]
```diff
@@ -423,6 +424,100 @@ case class ShowTablePropertiesCommand(
...
+      Seq.empty[Row]
+    } else {
+      val tab = catalog.getTable(table)
+      checkRequirements(tab)
+      val partNames = partitionSpec match {
+        case None => catalog.getPartitionNames(db, table.identifier, -1.asInstanceOf[Short])
+        case Some(spec) =>
+          validatePartitionSpec(tab, spec)
+          catalog.getPartitionNames(db, table.identifier, spec, -1.asInstanceOf[Short])
+      }
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: if you take my suggestion elsewhere then you don't need to do this match. Just pass the option into catalog.

Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: if you still want to validate the partition spec just do it in an option foreach or something

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala [hunk 10]
```diff
@@ -122,4 +133,103 @@ class HiveCommandSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
...
+
+      val message3 = intercept[AnalysisException] {
+        sql("SHOW PARTITIONS parquet_view1")
+      }.getMessage
+      assert(message3.contains("Operation not allowed: view or index table"))
+    }
+  }
+
+  test("show partitions - datasource") {
+    import sqlContext.implicits._
```
Comment (andrewor14) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: just `import testImplicits._` at the top

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala [hunk 11]
```diff
@@ -122,4 +133,103 @@ class HiveCommandSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
...
+      df.write
+        .partitionBy("a")
+        .format("parquet")
+        .mode(SaveMode.Overwrite)
+        .saveAsTable("part_datasrc")
+
+      val message1 = intercept[AnalysisException] {
+        sql("SHOW PARTITIONS part_datasrc")
+      }.getMessage
+      assert(message1.contains("Operation not allowed: datasource table"))
```
Comment (andrewor14) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: this assert is way too specific, just make it lower case and grep for "operation not allowed"

**Comment by andrewor14** [comment]
@dilipbiswal Thanks for working on this. I think it looks pretty good overall. The existing API for show partitions is a little too tied to Hive so I suggested an alternative to make it more consistent with the rest of the catalog. Other than that there weren't really any major issues.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 3 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: +1 we should follow current catalog API especially we already have `listPartitions` method.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala (hunk 11 above)
Comment (dilipbiswal) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: ok

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala (hunk 10 above)
Comment (dilipbiswal) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: ok

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 8 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Sure. Will make the change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 7 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: ok.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 5 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Will make the change

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 6 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: OK. will make the change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 4 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: ok.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 3 above)
Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Sure Andrew.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 2 above)
Comment (dilipbiswal) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Sorry, i should be passing the range to hive's API. I had it hard coded to -1 initially and later changed , so that caller of this API can pass in a range to restrict the number of partitions that are returned to the client. I will make the change to pass the supplied range to hive.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 4 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: @andrewor14 Hi Andrew, actually i remember i had tried this before. The table metadata's schema does not include partition columns ? Thats why i had used the lookupRelation.. 

Please let me know what you think ..

Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: it does now

Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: @andrewor14 Thanks Andrew. Let me re-base now :-)

**Comment by dilipbiswal** [comment]
@andrewor14 Thank you for a detailed review. I have implemented your comments. A few points/questions.
1) Currently InMemoryCatalog does not make use of partiton spec. If it is ok with you, can i work on it as a follow up ?
2) In listPartitions API , the optional partition spec has a default value to avoid changing the callers. Is that a ok thing to do ? If its not, i will change the callers to pass None where applicable.

Thanks again for your time and help !!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala [hunk 12]
```diff
@@ -277,9 +277,18 @@ class InMemoryCatalog extends ExternalCatalog {
     catalog(db).tables(table).partitions(spec)
   }
 
+  /**
+   * List the metadata of all partitions that belong to the specified table, assuming it exists.
+   *
+   * A partial partition spec may optionally be provided to filter the partitions returned.
+   * For instance, if there exist partitions (a='1', b='2'), (a='1', b='3') and (a='2', b='4'),
+   * then a partial spec of (a='1') will return the first two only.
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: Nit: No need to repeat the comment here since it's automatically inherited from the parent class/trait.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 13]
```diff
@@ -427,13 +427,18 @@ class SessionCatalog(
   }
 
   /**
-   * List all partitions in a table, assuming it exists.
-   * If no database is specified, assume the table is in the current database.
+   * List the metadata of all partitions that belong to the specified table, assuming it exists.
+   *
+   * A partial partition spec may optionally be provided to filter the partitions returned.
+   * For instance, if there exist partitions (a='1', b='2'), (a='1', b='3') and (a='2', b='4'),
+   * then a partial spec of (a='1') will return the first two only.
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Same as above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 14]
```diff
@@ -427,13 +427,18 @@ class SessionCatalog(
...
    */
-  def listPartitions(tableName: TableIdentifier): Seq[CatalogTablePartition] = {
+  def listPartitions(
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Nit: Missing `override`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala [hunk 15]
```diff
@@ -277,9 +277,18 @@ class InMemoryCatalog extends ExternalCatalog {
...
+   * TODO: Currently partialSpec is not used for memory catalog and it returns all the partitions.
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: Can we throw exception instead of returning all the partitions if a partition spec is given to an in-memory catalog? Silently returning wrong answer can be dangerous.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 16]
```diff
@@ -423,6 +426,102 @@ case class ShowTablePropertiesCommand(
 }
 
 /**
+ * A command for users to list the column names for a table. This function creates a
+ * [[ShowColumnsCommand]] logical plan.
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: "A command to list column names of a given table."

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 17]
```diff
@@ -423,6 +426,102 @@ case class ShowTablePropertiesCommand(
...
+ *
+ * The syntax of using this command in SQL is:
+ * {{{
+ *   SHOW COLUMNS (FROM | IN) table_identifier [(FROM | IN) database];
+ * }}}
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: We don't need to mention the SQL syntax here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 18]
```diff
@@ -423,6 +426,102 @@ case class ShowTablePropertiesCommand(
...
+case class ShowPartitionsCommand(
+    table: TableIdentifier,
+    spec: Option[TablePartitionSpec]) extends RunnableCommand {
+  // The result of SHOW PARTITIONS has one column called 'result'
+  override val output: Seq[Attribute] = {
+    AttributeReference("result", StringType, nullable = false)() :: Nil
+  }
+
+  def getPartName(spec: TablePartitionSpec): String = {
+    spec.map {s =>
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Space after `{`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 19]
```diff
@@ -423,6 +426,102 @@ case class ShowTablePropertiesCommand(
...
+      PartitioningUtils.escapePathName(s._1) + "=" + PartitioningUtils.escapePathName(s._2)
+    }.mkString("/")
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: I don't think we can simply join all parts using here since `TablePartitionSpec` is a `Map` rather than a `Seq`. Order of partition columns is not preserved.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 20]
```diff
@@ -423,6 +426,102 @@ case class ShowTablePropertiesCommand(
...
+ */
+case class ShowPartitionsCommand(
+    table: TableIdentifier,
+    spec: Option[TablePartitionSpec]) extends RunnableCommand {
+  // The result of SHOW PARTITIONS has one column called 'result'
+  override val output: Seq[Attribute] = {
+    AttributeReference("result", StringType, nullable = false)() :: Nil
+  }
+
+  def getPartName(spec: TablePartitionSpec): String = {
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Mark this as private.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 21]
```diff
@@ -423,6 +426,102 @@ case class ShowTablePropertiesCommand(
...
+    spec.map {s =>
+      PartitioningUtils.escapePathName(s._1) + "=" + PartitioningUtils.escapePathName(s._2)
+    }.mkString("/")
+  }
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: New line here please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 22]
```diff
@@ -423,6 +426,102 @@ case class ShowTablePropertiesCommand(
...
+  override def run(sqlContext: SQLContext): Seq[Row] = {
+    val catalog = sqlContext.sessionState.catalog
+    val db = table.database.getOrElse(catalog.getCurrentDatabase)
+    if (catalog.isTemporaryTable(table)) {
+      throw new AnalysisException("SHOW PARTITIONS is not allowed on a temporary table: " +
+          s"${table.unquotedString}")
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Wrong indentation here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 19 above)
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: @rxin I'm thinking it might not be a good idea to use `Map` to represent partition spec in `CatalogTablePartition` since it doesn't preserve partition column order. What do you think?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala [hunk 23]
```diff
@@ -122,4 +134,105 @@ class HiveCommandSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
...
+    assert(message.contains("badtable not found in database"))
+  }
+
+  test("show partitions - show everything") {
+    checkAnswer(
+      sql("show partitions parquet_tab4"),
+      Row("year=2015/month=1") ::
+        Row("year=2015/month=2") ::
+        Row("year=2016/month=2") ::
+        Row("year=2016/month=3") :: Nil)
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: Could you please add test cases where the testing table containing >= 5 partition columns? I believe Scala standard library provides specialized classes for maps containing <= 4 elements, which covers the out-of-order bug.

Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: As a simple experiment, order is preserved for maps with less than 5 elements:

```scala
scala> Map("year" -> 1) foreach println
(year,1)

scala> Map("year" -> 1, "month" -> 2) foreach println
(year,1)
(month,2)

scala> Map("year" -> 1, "month" -> 2, "day" -> 3) foreach println
(year,1)
(month,2)
(day,3)

scala> Map("year" -> 1, "month" -> 2, "day" -> 3, "hour" -> 4) foreach println
(year,1)
(month,2)
(day,3)
(hour,4)

scala> Map("year" -> 1, "month" -> 2, "day" -> 3, "hour" -> 4, "minute" -> 5) foreach println
(minute,5)
(year,1)
(hour,4)
(day,3)
(month,2)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 24]
```diff
@@ -423,6 +426,102 @@ case class ShowTablePropertiesCommand(
...
+      /**
+       * Validate the partitioning spec by making sure all the referenced columns are
+       * defined as partitioning columns in table definition. An AnalysisException exception is
+       * thrown if the partitioning spec is invalid.
+       */
+      if (spec.isDefined) {
+        val badColumns = spec.get.keySet.filterNot(tab.partitionColumns.map(_.name).contains)
+        if (badColumns.nonEmpty) {
+          throw new AnalysisException(
+            s"Non-partitioned column(s) [${badColumns.mkString(", ")}] are " +
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: Nit: Non-partitioning

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 25]
```diff
@@ -424,7 +424,7 @@ private[sql] object PartitioningUtils {
     path.foreach { c =>
       if (needsEscaping(c)) {
         builder.append('%')
-        builder.append(f"${c.asInstanceOf[Int]}%02x")
+        builder.append(f"${c.asInstanceOf[Int]}%02X")
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Why this change is necessary?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala [hunk 26]
```diff
@@ -17,11 +17,13 @@
...
-import org.apache.spark.sql.{AnalysisException, QueryTest, Row}
-import org.apache.spark.sql.hive.test.TestHiveSingleton
+import org.apache.spark.sql.{AnalysisException, QueryTest, Row, SaveMode}
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException
+import org.apache.spark.sql.hive.test.{TestHive, TestHiveSingleton}
 import org.apache.spark.sql.test.SQLTestUtils
 
 class HiveCommandSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
+  import testImplicits._
    protected override def beforeAll(): Unit = {
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: Please help fix this indentation, thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala [hunk 27]
```diff
@@ -36,12 +38,22 @@ class HiveCommandSuite extends QueryTest with SQLTestUtils with TestHiveSingleto
         |STORED AS PARQUET
         |TBLPROPERTIES('prop1Key'="prop1Val", '`prop2Key`'="prop2Val")
       """.stripMargin)
+     sql("CREATE TABLE parquet_tab3(col1 int, `col 2` int)")
+     sql("CREATE TABLE parquet_tab4 (price int, qty int) partitioned by (year int, month int)")
+     sql("INSERT INTO parquet_tab4 PARTITION(year = 2015, month=1) SELECT 1,1")
+     sql("INSERT INTO parquet_tab4 PARTITION(year = 2015, month=2) SELECT 2,2")
+     sql("INSERT INTO parquet_tab4 PARTITION(year = 2016, month=2) SELECT 3,3")
+     sql("INSERT INTO parquet_tab4 PARTITION(year = 2016, month=3) SELECT 3,3")
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: Nit: Spaces around `=`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala (hunk 12 above)
Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala]: Sure. I will remove the comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 16 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: ok.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala (hunk 17 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: @liancheng I followed other commands and some of them mention the SQL syntax. Do you want them removed from all other commands or just this two ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala (hunk 25 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: @liancheng So i was comparing the output of our impl against hive. And hive reports the escaped name in upper case,  I looked at FileUtils.escapePathName() in hive as reference before changing the code here. Please let me know what you think.

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Makes sense. Thanks for the explanation.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala (hunk 23 above)
Comment (dilipbiswal) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveCommandSuite.scala]: @liancheng Thanks !! You are right !! Having >= 5 partition keys does expose the problem. Any advice on how to go about handling this ? Can we change TablePartitionSpec to be a LinkedHashMap instead ?

**Comment by liancheng** [comment]
Hi @dilipbiswal, we did a bunch of major refactoring on the master branch, and it's pretty close to 2.0 code freeze, so I took it over based on your version and opened PR #12703. Would you mind to have a look at it? Thanks.

**Comment by dilipbiswal** [comment]
@liancheng I had all the comments addressed and had left the tests running last night was about to push today :-). Anyways thanks !! I will take a look at your PR.

**Comment by dilipbiswal** [comment]
@liancheng Hi Lian, i just pushed my branch which has the new test case for your reference. It also has all your comments addressed :-)

**Comment by liancheng** [comment]
@dilipbiswal Great! LGTM now, I'm merging this one to master. Thanks!

**Comment by dilipbiswal** [comment]
@liancheng Thank you very much Lian !!
